File: apps/php-sdk/src/Models/SearchOptions.php
<?php

declare(strict_types=1);

namespace Firecrawl\Models;

final class SearchOptions
{
    /**
     * @param list<mixed>|null $sources
     * @param list<mixed>|null $categories
     * @param list<string>|null $includeDomains
     * @param list<string>|null $excludeDomains
     */
    private function __construct(
        private readonly ?array $sources = null,
        private readonly ?array $categories = null,
        private readonly ?int $limit = null,
        private readonly ?string $tbs = null,
        private readonly ?string $location = null,
        private readonly ?bool $ignoreInvalidURLs = null,
        private readonly ?int $timeout = null,
        private readonly ?ScrapeOptions $scrapeOptions = null,
        private readonly ?string $integration = null,
        private readonly ?array $includeDomains = null,
        private readonly ?array $excludeDomains = null,
        private readonly ?bool $highlights = null,
    ) {}

    /**
     * @param list<mixed>|null $sources
     * @param list<mixed>|null $categories
     * @param list<string>|null $includeDomains
     * @param list<string>|null $excludeDomains
     */
    public static function with(
        ?array $sources = null,
        ?array $categories = null,
        ?int $limit = null,
        ?string $tbs = null,
        ?string $location = null,
        ?bool $ignoreInvalidURLs = null,
        ?int $timeout = null,
        ?ScrapeOptions $scrapeOptions = null,
        ?string $integration = null,
        ?array $includeDomains = null,
        ?array $excludeDomains = null,
        ?bool $highlights = null,
    ): self {
        return new self(
            $sources, $categories, $limit, $tbs, $location, $ignoreInvalidURLs,
            $timeout, $scrapeOptions, $integration, $includeDomains, $excludeDomains,
            $highlights,
        );
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        $fields = [
            'sources' => $this->sources,
            'categories' => $this->categories,
            'includeDomains' => $this->includeDomains,
            'excludeDomains' => $this->excludeDomains,
            'limit' => $this->limit,
            'tbs' => $this->tbs,
            'location' => $this->location,
            'ignoreInvalidURLs' => $this->ignoreInvalidURLs,
            'timeout' => $this->timeout,
            'highlights' => $this->highlights,
            'scrapeOptions' => $this->scrapeOptions?->toArray(),
            'integration' => $this->integration,
        ];

        return array_filter($fields, fn (mixed $v): bool => $v !== null);
    }
}

File: apps/.net-sdk/Firecrawl/Models/SearchOptions.cs
using System.Text.Json.Serialization;

namespace Firecrawl.Models;

/// <summary>
/// Configuration options for web search.
/// </summary>
public class SearchOptions
{
    [JsonPropertyName("sources")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<object>? Sources { get; set; }

    [JsonPropertyName("categories")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<object>? Categories { get; set; }

    [JsonPropertyName("limit")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Limit { get; set; }

    [JsonPropertyName("tbs")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Tbs { get; set; }

    [JsonPropertyName("location")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Location { get; set; }

    [JsonPropertyName("ignoreInvalidURLs")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? IgnoreInvalidURLs { get; set; }

    [JsonPropertyName("timeout")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Timeout { get; set; }

    [JsonPropertyName("scrapeOptions")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ScrapeOptions? ScrapeOptions { get; set; }

    [JsonPropertyName("integration")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Integration { get; set; }

    [JsonPropertyName("country")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Country { get; set; }

    [JsonPropertyName("enterprise")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? Enterprise { get; set; }
}
